docs(aws): dual-region setup in AWS (#3430)
camunda/camunda-docs: docs/self-managed/platform-deployment/helm-kubernetes/platforms/amazon-eks/dual-region.md

:::

## Outcome

Completion of this tutorial will result in:

- Two Amazon EKS Kubernetes clusters in two different geographic regions with each four nodes ready for the Camunda 8 installation.
- The [EBS CSI driver](https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html) installed and configured, which is used by the Camunda 8 Helm chart to create [persistent volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/).
- A [VPC peering](https://docs.aws.amazon.com/vpc/latest/peering/what-is-vpc-peering.html) between the two EKS clusters, allowing cross-cluster communication between different regions.
- An [Amazon Simple Storage Service](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html) (S3) bucket for [Elasticsearch backups](https://www.elastic.co/guide/en/elasticsearch/reference/current/repository-s3.html).

## Environment prerequisites

To streamline the execution of the subsequent commands, it is recommended to export multiple environment variables.

The following are the required environment variables with some example values:

Feel free to adjust the region and cluster names to your needs.

> ```bash

Author: why would you have to create a file?

I slightly adjusted it to explicitly mention the terminal.

```bash
git clone https://github.com/camunda/c8-multi-region.git
```

2. Navigate to `test/resources/aws/2-region/terraform`. This contains the Terraform base configuration for the dual-region setup.

### Contents elaboration

#### config.tf

This file contains the [backend](https://developer.hashicorp.com/terraform/language/settings/backends/configuration) and [provider](https://developer.hashicorp.com/terraform/language/providers/configuration) configuration.

Meaning where to store the [Terraform state](https://developer.hashicorp.com/terraform/language/state) and which providers to use, their versions, and potential credentials.

The important part of `config.tf` is the initialization of two AWS providers, as you need one per region and this is a limitation by AWS given everything is scoped to a region.

:::note

It's recommended to use a different backend than `local`. Find more information in the [Terraform documentation](https://developer.hashicorp.com/terraform/language/settings/backends/configuration).

:::

:::warning

Do not store sensitive information (credentials) in your Terraform files.

:::

> #### clusters.tf

Author: as mentioned in the message.
> I'm aware of branches still being wrong, blocked by missing review there + I'd move the current resources to root

This file covers the declaration of an S3 bucket to use for the backups. Additionally, a service account with access to use within the Kubernetes cluster to configure Elasticsearch to access the S3 bucket.

#### output.tf

[Terraform outputs](https://developer.hashicorp.com/terraform/language/values/outputs) allow you to reuse generated values in future steps. For example, the access keys of the service account with S3 access.

#### variables.tf

This file contains various variable definitions for both [local](https://developer.hashicorp.com/terraform/language/values/locals) and [input](https://developer.hashicorp.com/terraform/language/values/variables) types. The difference is that input variables require you to define the value on execution. While local variables are permanently defined, they are namely for code duplication purposes and readability.

### Preparation

1. Adjust any values in the `variables.tf` to your liking. For example, the target regions and their name or CIDR blocks of each cluster.
2. Make sure that any adjustments are reflected in your [environment prerequisites](#environment-prerequisites) to ease the [in-cluster setup](#in-cluster-setup).

> 3. Set up the authentication for the `AWS` provider. In the example, an AWS profile is used and provided as a Terraform variable that is defined on execution.

Author: I kept it as `Set up the authentication for the `AWS` provider` since we have the exact same thing in the other guides as well, also with the same disclaimer.

:::

### Execution

:::warning

A user who creates resources in AWS, will be their owner. In this particular case, the user will always have admin access to the Kubernetes cluster until the cluster is deleted.

Therefore, it can make sense to create an extra [AWS IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html) which credentials are used for Terraform purposes.

:::

1. Open a terminal and navigate to `test/resources/aws/2-region/terraform`.
2. Initialize the working directory:

```hcl
terraform init -upgrade
```

3. Apply the configuration files:

```hcl
terraform apply
```

If you have not set a default value for `cluster_name`, you will be asked to provide a suitable name.

4. After reviewing the plan, you can type `yes` to confirm and apply the changes.

At this point, Terraform will create the Amazon EKS clusters with all the necessary configurations. The completion of this process may require approximately 20-30 minutes.

## In-cluster setup

Now that you have created two Kubernetes clusters with Terraform, you will still have to configure various things to make the dual-region work.

> <Tabs queryString="cluster-setup">

Reviewer: See my comment about abother `<Tabs>` group & UX below.

# execute on cluster-paris
HOST2=$(kubectl --context $CLUSTER_1 -n kube-system get svc across-cluster-dns-tcp -o jsonpath="{.status.loadBalancer.ingress[0].hostname}" | cut -d - -f 1)
```

The following will print the IP addresses of the internal load-balancer in each cluster, which you will use to configure the opposite cluster for DNS resolving:

```bash
# gives you the IPs from region London that need to be configured in Paris
aws ec2 describe-network-interfaces --region $REGION_0 --filters Name=description,Values="ELB net/${HOST1}*" --query  'NetworkInterfaces[*].PrivateIpAddress' --output text
# Output example
# 10.190.114.39   10.190.172.21   10.190.134.196

# gives you the IPs from region Paris that need to be configured in London
aws ec2 describe-network-interfaces --region $REGION_1 --filters Name=description,Values="ELB net/${HOST2}*" --query  'NetworkInterfaces[*].PrivateIpAddress' --output text
# Output example
# 10.202.30.193   10.202.46.72   10.202.65.75
```

#### CoreDNS configuration

In the next part, you will configure CoreDNS to forward DNS queries for specific namespaces to the other cluster.

> You have to choose unique namespaces for Camunda 8 installations. The namespace for Camunda 8 installation in the cluster of region 0, needs to have a different name from the namespace for Camunda 8 installation in the cluster of region 1. This is required for proper traffic routing between the clusters.

Reviewer: ```suggestion
You have to choose unique namespaces for Camunda 8 installations. The namespace for Camunda 8 installation in the cluster of region 0 (`CAMUNDA_NAMESPACE_0`), needs to have a different name from the namespace for Camunda 8 installation in the cluster of region 1 (`CAMUNDA_NAMESPACE_1`). This is required for proper traffic routing between the clusters.
```

For example, you can install Camunda 8 into the `camunda-london` namespace in the `cluster-london` cluster, and `camunda-paris` namespace on the `cluster-paris` cluster.

> Using the same namespace on both clusters won't work as CoreDNS won't be able to distinguish between traffic targeted at the local and remote cluster.

Reviewer: ```suggestion
Using the same namespace names on both clusters won't work as CoreDNS won't be able to distinguish between traffic targeted at the local and remote cluster.
```
I know it's a bit of tautology, but just to ensure people understand it's not about not installing Camunda twice into a same namespace, but also about not having 2 samely named namespaces in different clusters.

        health {
            lameduck 5s
          }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
    camunda-paris.svc.cluster.local:53 {
        errors
        cache 30
        forward . 10.202.30.193 10.202.46.72 10.202.65.75 {
            force_tcp
        }
    }
    camunda-paris-failover.svc.cluster.local:53 {
        errors
        cache 30
        forward . 10.202.30.193 10.202.46.72 10.202.65.75 {
            force_tcp
        }
    }
```

  </summary>
</details>

2. Repeat **Step 1** for the Paris cluster to configure the CoreDNS of Paris to redirect any DNS requests to the London DNS resolver for your chosen namespace, like `camunda-london`.

3. Check that CoreDNS has reloaded for the changes to take effect before continuing. Make sure it contains `Reloading complete`:

```bash
kubectl --context $CLUSTER_0 logs -f deployment/coredns -n kube-system
kubectl --context $CLUSTER_1 logs -f deployment/coredns -n kube-system
```

</TabItem>

> <TabItem value="test-dns-chaining" label="(optional) Test DNS chaining">

Reviewer: let's drop `optional` here, I think it would be better the users check their setup always, as early in the process as possible.

A setup to test that the DNS chaining is working by using nginx pods and services to ping each other.

1. Deploy the following nginx pod and service to each cluster within the `camunda-london` and `camunda-paris` namespaces:

```bash
kubectl --context $CLUSTER_0 apply -f https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/kubernetes/nginx.yml -n $CAMUNDA_NAMESPACE_0
kubectl --context $CLUSTER_1 apply -f https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/kubernetes/nginx.yml -n $CAMUNDA_NAMESPACE_1
```

> 2. Try to curl each other from within the other pod:

Reviewer: ```suggestion
2. Try to `curl` each other from within the other pod:
```

```bash
kubectl --context $CLUSTER_0 exec -n $CAMUNDA_NAMESPACE_0 -it sample-nginx -- curl http://sample-nginx.sample-nginx-peer.$CAMUNDA_NAMESPACE_0.svc.cluster.local

kubectl --context $CLUSTER_1 exec -n $CAMUNDA_NAMESPACE_1 -it sample-nginx -- curl http://sample-nginx.sample-nginx-peer.$CAMUNDA_NAMESPACE_1.svc.cluster.local
```

3. After a successful retrieval of the remote nginx start page, you can remove this temporary setup.

```bash
kubectl --context $CLUSTER_0 delete -f https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/kubernetes/nginx.yml -n $CAMUNDA_NAMESPACE_0
kubectl --context $CLUSTER_1 delete -f https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/kubernetes/nginx.yml -n $CAMUNDA_NAMESPACE_1
```

</TabItem>
</Tabs>

## Deploy Camunda 8 to the clusters

> <Tabs queryString="deploy">

Reviewer: Same as for previous tabs incarnation, it's not obvious that a user needs to go through steps one by one.
I suggest:
- adding `Step <N>. ` prefix to every tab name
- at the end of every tab content, add an explicit text to go to the next tab `Step <N+1>. `

Author: I'm aware of this, I wanted to first explore whether we keep the tabs or not, now that the general procedure is shorter due to the switch to cloning the repo. As now we have a part without tabs and the rest with tabs.

Nevertheless, good suggestion with the step if we keep the tabs.

<TabItem value="elastic-secret" label="Create the secret for Elasticsearch">

> Elasticsearch will use S3 as a backup and restore bucket. For this, configure a Kubernetes secret to not expose those in cleartext.

Reviewer: ```suggestion
Elasticsearch will need an S3 bucket for data backup and restore procedure, required during a regional failover. For this, you will need to configure a Kubernetes secret to not expose those in cleartext:
```

You can pull the data from Terraform since you exposed those via the `output.tf`.

1. From the Terraform code location `test/resources/aws/2-region/terraform`, execute the following to export the access keys to environment variables. This will allow an easier creation of the Kubernetes secret via the command line:

```bash
export ACCESS_KEY=$(terraform output -raw s3_aws_access_key_id)
export SECRET_ACCESS_KEY=$(terraform output -raw s3_aws_secret_access_key)
```

2. Pre-create the Camunda namespaces in each cluster. This is required to have the secrets available for Elasticsearch before installing the Helm chart.

> ```bash

Reviewer: Does it make sense to move this whole step into a small shell script that creates the secrets?

kubectl --context $CLUSTER_1 create namespace $CAMUNDA_NAMESPACE_1_FAILOVER
```

3. Create the Kubernetes secrets in each of the clusters, where Camunda will be installed, including those of the failover

```bash
kubectl --context $CLUSTER_0 -n $CAMUNDA_NAMESPACE_0 create secret generic elasticsearch-env-secret \
    --from-literal=S3_ACCESS_KEY=$ACCESS_KEY \
    --from-literal=S3_SECRET_KEY=$SECRET_ACCESS_KEY

kubectl --context $CLUSTER_0 -n $CAMUNDA_NAMESPACE_0_FAILOVER create secret generic elasticsearch-env-secret \
    --from-literal=S3_ACCESS_KEY=$ACCESS_KEY \
    --from-literal=S3_SECRET_KEY=$SECRET_ACCESS_KEY

kubectl --context $CLUSTER_1 -n $CAMUNDA_NAMESPACE_1 create secret generic elasticsearch-env-secret \
    --from-literal=S3_ACCESS_KEY=$ACCESS_KEY \
    --from-literal=S3_SECRET_KEY=$SECRET_ACCESS_KEY

kubectl --context $CLUSTER_1 -n $CAMUNDA_NAMESPACE_1_FAILOVER create secret generic elasticsearch-env-secret \
    --from-literal=S3_ACCESS_KEY=$ACCESS_KEY \
    --from-literal=S3_SECRET_KEY=$SECRET_ACCESS_KEY
```

3. Unset environment variables to reduce the risk of potential exposure.

```bash
unset ACCESS_KEY

> unset SECRET_ACCESS_KEYv

Reviewer: ```suggestion
unset SECRET_ACCESS_KEY
```

```

</TabItem>
<TabItem value="helm-chart" label="Camunda 8 Helm chart prerequisites">

> 1. Within the cloned repository, navigate to `test/resources/aws/2-region/kubernetes`. This contains a dual-region example setup.

Reviewer: Reminder to rename when it happens.

#### Preparation

The base `camunda-values.yml` requires some adjustments before installing the Helm chart.

1. Adjust the `placeholder` for `ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS` to reflect your current setup. These are the contact points for the brokers to know how to form the cluster. Find more information on what the variable means in [setting up a cluster](../../../../zeebe-deployment/operations/setting-up-a-cluster.md).

<!-- TODO: currently on a different branch - infex-98-docs-changes -->

**Use the python script in the root of the repository `generate_zeebe_initial_contact.py` to generate the replacement string.**

```bash
python3 generate_zeebe_initial_contact.py

# It will ask you to provide the following values
# Enter a namespace for cluster in region 0:
## in our case camunda-london
# Enter a namespace for cluster in region 1:
## in our case camunda-paris
# Enter Helm release name:
## the way you'll call the Helm release, for example camunda
# Enter Zeebe cluster size:
## for a dual-region setup we recommend 8
```

In our example, the `placeholder` is replaced with the following:

> ```yaml

Reviewer: Let's make a more clear explanation that the users should not use these values, they are just for illustration.

camunda-zeebe-0.camunda-zeebe.camunda-london.svc.cluster.local:26502,camunda-zeebe-1.camunda-zeebe.camunda-london.svc.cluster.local:26502,camunda-zeebe-2.camunda-zeebe.camunda-london.svc.cluster.local:26502,camunda-zeebe-3.camunda-zeebe.camunda-london.svc.cluster.local:26502,camunda-zeebe-0.camunda-zeebe.camunda-paris.svc.cluster.local:26502,camunda-zeebe-1.camunda-zeebe.camunda-paris.svc.cluster.local:26502,camunda-zeebe-2.camunda-zeebe.camunda-paris.svc.cluster.local:26502,camunda-zeebe-3.camunda-zeebe.camunda-paris.svc.cluster.local:26502
```

The Zeebe cluster size is eight in a dual-region setup, resulting in four brokers per region.

2. In case the Helm release name differs from `camunda` and your chosen namespaces, you must adjust the `ZEEBE_BROKER_EXPORTERS_ELASTICSEARCHREGION0_ARGS_URL` and `ZEEBE_BROKER_EXPORTERS_ELASTICSEARCHREGION1_ARGS_URL` values.

They are constructed as follows:

```bash
http://${HELM_RELASE_NAME}-elasticsearch-master-hl.${CAMUNDA_NAMESPACE}.svc.cluster.local:9200
# Where HELM_RELEASE_NAME is your chosen helm release name

> # and CAMUNDA_NAMESPACE is your previously defined CAMUNDA_NAMESPACE_0 and CAMUNDA_NAMESPACE_1, depending on ELASTICSEARCHREGION0 or ELASTICSEARCHREGION1

Reviewer: I'm confused about this part, what does `depending on ELASTICSEARCHREGION0 or ELASTICSEARCHREGION1` mean?

`ELASTICSEARCHREGION0` and `ELASTICSEARCHREGION1` have not been previously mentioned.

Author: rephrased it and added an example.

```

</TabItem>
<TabItem value="deploy" label="Deploy Camunda 8">

> The last step is to deploy Camunda 8 to both its regions, for which we chose London as `Region 0` and Paris as `Region 1`. As mentioned, you may choose a different release name and should ensure you have adjusted your base `camunda-values.yml` to consider this:

Reviewer: ```suggestion
The last step is to deploy Camunda 8 to both its regions `REGION_0` and `REGION_1`. As mentioned, you may choose a different release name and should ensure you have adjusted your base `camunda-values.yml` to consider this:
```

Reviewer: > should ensure you have adjusted your base `camunda-values.yml` to consider this:

not sure what this means, what exactly do I need to change in that yaml file?

Author: rephrased it

1. From the terminal context of `test/resources/aws/2-region/kubernetes` execute:

```bash
helm install camunda camunda/camunda-platform \
  --version 9.3.0 \
  --kube-context $CLUSTER_0 \
  --namespace $CAMUNDA_NAMESPACE_0 \

>   -f base-values.yml \

Reviewer: there's no `base-values.yml` in the git repo.

>   -f region0-values.yml

Reviewer: there's no `region0-values.yml` in the git repo.

- A [VPC peering](https://docs.aws.amazon.com/vpc/latest/peering/what-is-vpc-peering.html) between the two EKS clusters, allowing cross-cluster communication between different regions.
- An [Amazon Simple Storage Service](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html) (S3) bucket for [ElasticSearch backups](https://www.elastic.co/guide/en/elasticsearch/reference/current/repository-s3.html)

## Environment prerequisites

To streamline the execution of the subsequent commands, it is recommended to export multiple environment variables.

The following are the required environment variables with some example values:

```bash
# The AWS regions of your cluster 1 and 2
export REGION_1=eu-west-2
export REGION_2=eu-west-3

# The cluster_name of your cluster 1 and 2
# default based on the tutorial is the following
export CLUSTER_1=cluster-london
export CLUSTER_2=cluster-paris

# The namespaces for each region where Camunda 8 should be running in
export CAMUNDA_NAMESPACE_1=camunda-london
export CAMUNDA_NAMESPACE_2=camunda-paris
```

> ## Installing Amazon EKS clusters with Terraform

Reviewer: ## **[big one]**

I have a general feeling it's too much work for us to guide the user what files to create and what content to put there.

Maybe we can say the user needs to clone https://github.com/camunda/c8-multi-region/tree/aws-operational/test/resources/aws/2-region/terraform (with all files pre-created) and customize the Terraform variables and then just `tf apply`?

It's cool we can embed the code content, but I also have a feeling it won't be much useful for the user and create a lot of work for us to maintain/create such content.

wdyt?

Author: the proposed change is included in the latest commit - https://github.com/camunda/camunda-docs/pull/3430/commits/4e11aa0a397cdf49f8810c07b254f8420fb23e58

<Tabs queryString="tf-installation">
<TabItem value="terraform-prerequsites" label="Terraform prerequsites" default>

1. Create an empty folder to place your Terraform files in.
2. Create a `config.tf` with the following setup:

<details>
  <summary>config.tf</summary>
  <summary>

```hcl reference title="config.tf"
https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/terraform/config.tf#L5-L51
```

>   </summary>

Author: will keep it open for now, not sure whether still needed with the current suggestions.
Never noticed since I always open links with the scroll wheel press.

>   </summary>

Reviewer: Agree. Can be a bug of the Docusaurus plugging where they did not think of the corner case of when a code block is inside a collapsing block.

3. Feel free to adjust the local values to fit your configuration. E.g. different region or cidr blocks.
4. Set up the authentication for the `AWS` provider. In the example, an AWS profile is used and provided as a terraform variable that is defined on execution.

:::note

It's recommended to use a different backend than `local`. More information can be found in the [Terraform documentation](https://developer.hashicorp.com/terraform/language/settings/backends/configuration).

:::

:::note

The [AWS Terraform provider](https://registry.terraform.io/providers/hashicorp/aws/latest/docs) is required to create resources in AWS. You must configure the provider with the proper credentials before using it. You can further change the region and other preferences and explore different [authentication](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#authentication-and-configuration) methods.

There are several ways to authenticate the `AWS` provider.

- (Recommended) Use the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html) to configure access. Terraform will automatically default to AWS CLI configuration when present.

> - Set environment variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`, where the `key` and `id` can be retrieved from the [AWS Console](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html).

Author: adjusted it also in the 8.4 docs for the terraform guide.

:::

:::warning

> Do not use secrets in your configuration files.

Author: adjusted it for the 8.4 terraform guide as well.

:::

:::warning

> The user who creates the resources will always be the owner. This means the user will always have admin access to the Kubernetes cluster until you delete it. Therefore, it can make sense to create an extra [AWS IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html) that's solely used for Terraform purposes.

Author: adjusted it for the 8.4 terraform guide as well.

> :::

Author: thought about that as well already to have a `next step` link switching the tab.
Yes Deeplinks work

> :::

Author: keeping open for now, have to see how I approach this.
The querystring stuff will require a reload of the page and does not jump to the position.

The alternative is to simulate a click with javascript when clicking on a link.

> :::

Author: tabs removed

</TabItem>
<TabItem value="cluster-module" label="Cluster module">

This module creates the basic layout that configures AWS access and Terraform.

The following will use [Terraform modules](https://developer.hashicorp.com/terraform/language/modules), which allows abstracting resources into reusable components.

The [Camunda provided module](https://github.com/camunda/camunda-tf-eks-module) is publicly available. It's advisable to review this module before usage.

1. In the folder where your `config.tf` resides, create a file called `cluster.tf`.
2. Paste the following content into the created `cluster.tf` file to make use of the provided module:

<details>
  <summary>cluster.tf</summary>
  <summary>

```hcl reference title="cluster.tf"

> https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/terraform/config.tf#L57-L86

Author: the version is a linting thing, we can disable it.

# Output example
# 10.190.114.39   10.190.172.21   10.190.134.196

# gives you the IPs from region Paris that need to be configured in London
aws ec2 describe-network-interfaces --region $REGION_2 --filters Name=description,Values="ELB net/${HOST2}*" --query  'NetworkInterfaces[*].PrivateIpAddress' --output text
# Output example
# 10.202.30.193   10.202.46.72   10.202.65.75
```

Configure CoreDNS to forward DNS queries for specific namespaces to the other cluster.
You have to choose unique namespaces for Camunda 8 that only exist once in each cluster to direct the traffic.

For example, `camunda-london` on the `cluster-london` and `camunda-paris` on the `cluster-paris`.
Using the same namespace on both clusters won't work since CoreDNS won't be able to distinguish between traffic targeted at the local and remote cluster.

Besides the main namespaces, you will have to add `failover` namespaces for the unfortunate event of a total region loss. This is for completeness, so you don't forget to add the mapping on region recovery. The operational procedure is handled in a different [document](#). <!-- TOOD: add reference -->

> 

Author: good point, included since all namespaces have to be pre-created for the elastic secret.

3. You need to configure CoreDNS with the namespace to DNS resolver mapping.

```bash
kubectl --context $CLUSTER_1 edit -n kube-system configmap coredns
```

Add at the bottom of the `Corefile` the namespace mapping to the remote DNS resolver.

<!-- yaml is always reformated wrongly -->

```bash
# You are configuring the CoreDNS of the london cluster
# to redirect any DNS requests for camunda-paris to the paris cluster
camunda-paris.svc.cluster.local:53 {
  errors
  cache 30

>   forward . 10.202.30.193 10.202.46.72 10.202.65.75 {

Author: added an admonitions above it to cover the following 3 comments.

tl;dr
- Just example values
- Make sure to edit
- Won't work in your setup

    force_tcp
  }
}
camunda-paris-failover.svc.cluster.local:53 {
  errors
  cache 30
  forward . 10.202.30.193 10.202.46.72 10.202.65.75 {
    force_tcp
  }
}
```

<details>
  <summary>Full example</summary>
  <summary>

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    eks.amazonaws.com/component: coredns
    k8s-app: kube-dns
  name: coredns
  namespace: kube-system
data:
  Corefile: |
    .:53 {
        errors
        health {
            lameduck 5s
          }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }

>     camunda-paris.svc.cluster.local:53 {

Author: same as above already

            force_tcp
        }
    }
    camunda-paris-failover.svc.cluster.local:53 {
        errors
        cache 30
        forward . 10.202.30.193 10.202.46.72 10.202.65.75 {
            force_tcp
        }
    }
```

  </summary>
</details>

4. Repeat step 3 for the Paris cluster to configure the CoreDNS of Paris to redirect any DNS requests to the London DNS resolver for your chosen namespace, e.g. `camunda-london`.

5. Check that CoreDNS has reloaded for the changes to take effect before continuing. Make sure it contains `Reloading complete`.

```bash
kubectl --context $CLUSTER_1 logs -f deployment/coredns -n kube-system
kubectl --context $CLUSTER_2 logs -f deployment/coredns -n kube-system
```

</TabItem>
<TabItem value="test-dns-chaining" label="(optional) Test DNS chaining">

A simple setup to test that the DNS chaining is working by using nginx pods and services to ping each other.

1. Deploy the following nginx pod and service to each cluster within the `camunda-london` and `camunda-paris` namespaces.

```yaml reference title="nginx.yml"
https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/kubernetes/nginx.yml
```

```bash

> kubectl --context $CLUSTER_1 apply -f https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/kubernetes/nginx.yml -n $CAMUNDA_NAMESPACE_1

Author: I'm aware of that, currently blocked on that part.

kubectl --context $CLUSTER_2 apply -f https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/kubernetes/nginx.yml -n $CAMUNDA_NAMESPACE_2
```

2. Try to curl each other from within the other pod.

```bash
kubectl --context $CLUSTER_1 exec -n $CAMUNDA_NAMESPACE_1 -it sample-nginx -- curl http://sample-nginx.sample-nginx-peer.camunda-paris.svc.cluster.local

kubectl --context $CLUSTER_2 exec -n $CAMUNDA_NAMESPACE_2 -it sample-nginx -- curl http://sample-nginx.sample-nginx-peer.camunda-london.svc.cluster.local

> # http://sample-nginx.sample-nginx-peer.camunda-london.svc.cluster.local

Author: not needed, was just to showcase how it's constructed.

kubectl --context $CLUSTER_2 delete -f https://github.com/camunda/c8-multi-region/blob/main/test/resources/aws/2-region/kubernetes/nginx.yml -n $CAMUNDA_NAMESPACE_2
```

</TabItem>
</Tabs>

## Deploy Camunda 8 to the Clusters

<Tabs queryString="deploy">
<TabItem value="elastic-secret" label="Create the secret for ElasticSearch">

ElasticSearch will use S3 as a backup and restore bucket and for that, you need to configure a Kubernetes secret to not expose those in cleartext.

You can pull the data from Terraform since you exposed those via the `output.tf`.

1. Where your `config.tf` and `output.tf` live, execute the following to export the access keys to environment variables. This will allow an easier creation of the Kubernetes secret via the command line.

```bash
export ACCESS_KEY=$(terraform output -raw s3_aws_access_key_id)
export SECRET_ACCESS_KEY=$(terraform output -raw s3_aws_secret_access_key)
```

2. Create the Kubernetes secret within the `camunda-london` and `camunda-paris` namespace on each respective cluster.

```bash

> kubectl --context $CLUSTER_1 -n $CAMUNDA_NAMESPACE_1 create secret generic elasticsearch-env-secret \

Author: good catch

    --from-literal=S3_ACCESS_KEY=$ACCESS_KEY \
    --from-literal=S3_SECRET_KEY=$SECRET_ACCESS_KEY

kubectl --context $CLUSTER_2 -n $CAMUNDA_NAMESPACE_2 create secret generic elasticsearch-env-secret \
    --from-literal=S3_ACCESS_KEY=$ACCESS_KEY \
    --from-literal=S3_SECRET_KEY=$SECRET_ACCESS_KEY
```

</TabItem>
<TabItem value="helm-chart" label="Camunda 8 Helm Chart prerequisites">

> Our recommendation is to work with layered helm values files since you will want to have a base `value.yml` that is generally applicable and then two overlays that are for `region0` and `region1`.

Reviewer: ## [big one]

Same suggestion as for the TF files - does it make sense to instead to the users to `git clone` https://github.com/camunda/c8-multi-region/tree/aws-operational/test/resources/aws/2-region/kubernetes, adjust the scripts accordingly, so we don't need to explain what files with what contents to create?

Author: the proposed change is included in the latest commit - https://github.com/camunda/camunda-docs/pull/3430/commits/4e11aa0a397cdf49f8810c07b254f8420fb23e58

#### Base values.yml

The following is an example `values.yml` that forms the base and ca be further extended.

> 

Author: let me know whether we still have to iterate on this after the updates.

> Especially important within this is the `global.multiregion.regions` directive to be two for two regions. Additionally, to disable identity and optimize since those are currently not working within the setup.

Author: is part of that section

You need to adjust the `placeholder` for `ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS` to reflect your current setup. These are the contact points for the brokers to know how to form the cluster. More information on what the variable means in the [setting up a cluster documentation](../../../../zeebe-deployment/operations/setting-up-a-cluster.md).

In our example, the `placeholder` would is being replaced with:

```yaml
-name: ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS
 value: "camunda-zeebe-0.camunda-zeebe.camunda-london.svc.cluster.local:26502,camunda-zeebe-1.camunda-zeebe.camunda-london.svc.cluster.local:26502,camunda-zeebe-2.camunda-zeebe.camunda-london.svc.cluster.local:26502,camunda-zeebe-3.camunda-zeebe.camunda-london.svc.cluster.local:26502,camunda-zeebe-0.camunda-zeebe.camunda-paris.svc.cluster.local:26502,camunda-zeebe-1.camunda-zeebe.camunda-paris.svc.cluster.local:26502,camunda-zeebe-2.camunda-zeebe.camunda-paris.svc.cluster.local:26502,camunda-zeebe-3.camunda-zeebe.camunda-paris.svc.cluster.local:26502"
```

The zeebe cluster size is 8 with a dual-region setup, resulting in 4 brokers per region.

The address forms as follows:

```bash

> $pod.$service.$namespace.svc.cluster.local:26502

Reviewer: can something like https://github.com/camunda-community-hub/camunda-8-helm-profiles/blob/main/google/multi-region/active-active/setup-zeebe.py be used as a endpoint hostnames list generator?

i.e. a script put in https://github.com/camunda/c8-multi-region/tree/aws-operational/test/resources/aws/2-region/terraform and asking some basic inputs from a user / env vars (namespace name, broker count ...). It's a bit hard to generate those by hand without mistakes.

Author: yes, we can probably have a small python script.

Reviewer: a Python script to run would be yet another thing easier with the `git clone` approach should we go for it.

Author: was handled in - https://github.com/camunda/c8-multi-region/pull/17

```

- `$pod` = `$release`-zeebe-`$number`
  - `$release` is the helm release name that you use. In our case, we're going with `camunda`.
  - `$number` stems from zeebe being a statefulset, so it starts from 0 to 3 due to 4 replicas per region.
  - resulting in `camunda-zeebe-(0-4)`
- `$service` = `$release-zeebe`
  - `$release` is the helm release name that you use. In our case, we're going with `camunda`.
  - resulting in `camunda-release`
- `$namespace` is the previously created namespace, in our case `camunda-london` and `camunda-paris`.
- 26502 is the port for the initial contact at zeebe.

In case you are going with a different helm release name than `camunda`, you must adjust the URLs for the ElasticSearch exporters `ZEEBE_BROKER_EXPORTERS_ELASTICSEARCHREGION0_ARGS_URL` and `ZEEBE_BROKER_EXPORTERS_ELASTICSEARCHREGION1_ARGS_URL`. Where `camunda` must be replaced with the release name that you have chosen.

<details>
  <summary>base-values.yml</summary>
  <summary>

```yml reference title="base-values.yml"

> https://github.com/camunda/c8-multi-region/blob/aws-operational/test/resources/aws/2-region/kubernetes/camunda-values.yml

Author: from the secret `extraEnvVarsSecret`?

```

  </summary>
</details>

#### Region0 values.yml

The vital part of the overlay is that the `regionId` is set to 0. The rest is being handled in the base layer.

```yaml reference title="region0-values.yml"
https://github.com/camunda/c8-multi-region/blob/aws-operational/test/resources/aws/2-region/kubernetes/region0/camunda-values.yml
```

#### Region1 values.yml

The vital part of the overlay is that the `regionId` is set to 1. The rest is being handled in the base layer.

```yaml reference title="region1-values.yml"
https://github.com/camunda/c8-multi-region/blob/aws-operational/test/resources/aws/2-region/kubernetes/region1/camunda-values.yml
```

</TabItem>
<TabItem value="deploy" label="Deploy Camunda 8">

The last step is to deploy Camunda 8 to both its regions, for which we chose London as region 0 and Paris as region 1. As mentioned, you may choose a different release name and should ensure you have adjusted your `base-values.yml` to consider this.

```bash

> helm install camunda camunda/camunda-platform \

Author: we can pin it to 9.x but should generally work with any version moving forward.

> helm install camunda camunda/camunda-platform \

Author: should we give version recommendations?
Starting 8.3.x, 8.4.x, 9.x?

> helm install camunda camunda/camunda-platform \

Reviewer: maybe something to check with the Distribution team on how to refer to a Helm chart for 8.5.x stack. For now guess we can just say whatever Helm chart version will be there with 8.5.0, wdyt?

> helm install camunda camunda/camunda-platform \

Author: 8.5.x stack --> 9x+

> helm install camunda camunda/camunda-platform \

Author: ah you mean that, since it's already supported why not simply try to give a version recommendation?
We can go with the current 9.2 but maybe should have a matrix from when on it's supported.
8.3.x, 8.4.x, 9.x.y

> helm install camunda camunda/camunda-platform \

Reviewer: As we're not going to publicly advertise 8.3/8.4 support, let's use the Helm chart that refers to Camunda 8.5 y default (not sure which one will that be, but at least 9.x I guess)

  --kube-context $CLUSTER_1 \
  --namespace $CAMUNDA_NAMESPACE_1 \
  -f base-values.yml \
  -f region0-values.yml

helm install camunda camunda/camunda-platform \
  --kube-context $CLUSTER_2 \
  --namespace $CAMUNDA_NAMESPACE_2 \
  -f base-values.yml \
  -f region1-values.yml
```

> </TabItem>

Author: We can. I think it would be great to have a generally applicable getting started page since we keep repeating ourselves here.

> </TabItem>

Reviewer: I was thinking about it as well. Maybe this page has EKS-specific stuff and the C8 install in the multi-region mode moved out to a separate page (and is cloud-agnostic)?

> </TabItem>

Author: it's intertwined with AWS too much due to Elasticsearch.
What I mean is a generally applicable page on how to get started with Camunda 8, including verifying that it works.

> </TabItem>

Author: I think it's out of scope for this PR but something in the future to have a getting started guide with C8, might even exist already, who knows.

> </TabItem>

Author: I included a verify step with `zbctl`.
But as mentioned, I think we need a general page in the future, outside of current scope, that deals with this.

:::

## Outcome

Completion of this tutorial will result in:

- Two Amazon EKS Kubernetes clusters in two different geographic regions with each four nodes ready for the Camunda 8 installation.
- The [EBS CSI driver](https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html) installed and configured, which is used by the Camunda 8 Helm chart to create [persistent volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/).
- A [VPC peering](https://docs.aws.amazon.com/vpc/latest/peering/what-is-vpc-peering.html) between the two EKS clusters, allowing cross-cluster communication between different regions.
- An [Amazon Simple Storage Service](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html) (S3) bucket for [Elasticsearch backups](https://www.elastic.co/guide/en/elasticsearch/reference/current/repository-s3.html).

## Environment prerequisites

To streamline the execution of the subsequent commands, it is recommended to export multiple environment variables.

The following are the required environment variables with some example values:

Feel free to adjust the region and cluster names to your needs.

```bash

> # The AWS regions of your cluster 1 and 2

Reviewer: shall this read cluster 0 and cluster 1?

:::note

It's recommended to use a different backend than `local`. Find more information in the [Terraform documentation](https://developer.hashicorp.com/terraform/language/settings/backends/configuration).

:::

:::warning

Do not store sensitive information (credentials) in your Terraform files.

:::

#### clusters.tf

This file is using [Terraform modules](https://developer.hashicorp.com/terraform/language/modules), which allows abstracting resources into reusable components.

The [Camunda provided module](https://github.com/camunda/camunda-tf-eks-module) is publicly available. It's advisable to review this module before usage.

There are various other input options to customize the cluster setup further. See the [module documentation](https://github.com/camunda/camunda-tf-eks-module) for additional details.

It contains the declaration of the two clusters. One of them has an explicit provider declaration as otherwise everything is deployed to the default AWS provider, which is limited to a single region.

#### vpc-peering.tf

> For a multi-region setup, you need to have the [virtual private cloud (VPC)](https://aws.amazon.com/vpc/) peered to route traffic between them using private IPv4 addresses and not publicly route the traffic and expose it. For more information, review the [AWS documentation on VPC peering](https://docs.aws.amazon.com/vpc/latest/peering/what-is-vpc-peering.html).

Reviewer: ```suggestion
For a multi-region setup, you need to have the [virtual private cloud (VPC)](https://aws.amazon.com/vpc/) peered to route traffic between regions using private IPv4 addresses and not publicly route the traffic and expose it. For more information, review the [AWS documentation on VPC peering](https://docs.aws.amazon.com/vpc/latest/peering/what-is-vpc-peering.html).
```

```bash
unset ACCESS_KEY
unset SECRET_ACCESS_KEY
```

### Camunda 8 Helm chart prerequisites

Within the cloned repository, navigate to `test/resources/aws/2-region/kubernetes`. This contains a dual-region example setup.

#### Content Elaboration

Our approach is to work with layered helm values files:

- have a base `camunda-values.yml` that is generally applicable for both Camunda installations
- two overlays that are for region 0 and region 1 installations

##### camunda-values.yml

This forms the base layer that contains the basic required setup, which applies to both regions.

Key changes of the dual-region setup:

- `global.multiregion.regions: 2`
  - indicates the use for two regions
- `global.identity.auth.enabled: false`
  - Identity is currently not supported. Please see the [limitations section](#) on the dual-region concept page. <!-- TODO: fill link -->.
- `global.elasticsearch.disableExporter: true`
  - disables the automatic Elasticsearch configuration of the helm chart. We will manually supply the values via environment variables.
- `identity.enabled: false`
  - Identity is currently not supported.
- `optimize.enabled: false`

>   - Optimize is currently not supported. It has a dependency on identity.

Reviewer: ```suggestion
  - Optimize is currently not supported. It has a dependency on Identity.
```

- `zeebe.env`
  - `ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS`
    - These are the contact points for the brokers to know how to form the cluster. Find more information on what the variable means in [setting up a cluster](../../../../zeebe-deployment/operations/setting-up-a-cluster.md).
  - `ZEEBE_BROKER_EXPORTERS_ELASTICSEARCHREGION0_ARGS_URL`
    - The Elasticsearch endpoint for region 0.
  - `ZEEBE_BROKER_EXPORTERS_ELASTICSEARCHREGION1_ARGS_URL`
    - The Elasticsearch endpoint for region 1.

> - Following cluster of 8 is recommend for a dual-region setup

Reviewer: ```suggestion
- A cluster of 8 Zeebe brokers (4 in each of the regions) is recommended for the dual-region setup
```

>   - `zeebe.clusterSize: 8`

Reviewer: did you double check with Consulting/Zeebe that 8/8/8 is the correct setup? I had a vague memory of the hack week that not all of these values are 8, but I could be wrong.

Author: you're right, sorry.
it's 8/8/4, my `zbctl status` shows the correct one.
